### PR TITLE
Don't generate a SDK for runtime-versioner patching

### PR DESCRIPTION
**Pull request by jdisanti**
This PR removes the requirement of generating a SDK in order to patch smithy-rs runtime crates into an already released SDK version for testing. This is possible now that all the runtime crates are independently versioned. The tool also takes multiple paths as input for the runtime crate locations. This makes it possible to pass in the `rust-runtime` and `aws/rust-runtime` directories from smithy-rs (done by default by one of the subcommands).

Some of the publisher tool's package resolution code is moved into smithy-rs-tool-common so it can be better shared across the tools, and as part of this, the version in the package handle was made optional.

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

**Review by rcoh — APPROVED** [review]
consider removing the asserts by creating "UnpublishablePackage". We can keep version required except in the case where we need it to be optional.

My branch kinda went the wrong direction and make optional the default generic, but I think if we went the other direction and made non-optional verions the default we could remove a lot of asserts & make only small changes to the code.

File: tools/ci-build/smithy-rs-tool-common/src/package.rs [hunk 1]
```diff
@@ -46,3 +55,291 @@ pub enum PackageStability {
...
+pub enum Publish {
+    Allowed,
+    NotAllowed,
+}
+
+/// Represents a crate (called Package since crate is a reserved word).
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct Package {
+    /// Package name and version information
+    pub handle: PackageHandle,
```
Comment (aajtodd) [inline tools/ci-build/smithy-rs-tool-common/src/package.rs:108]: If it's just name, version, etc this feels more like `meta`, perhaps it should be `PackageMeta`

Comment (jdisanti) [reply tools/ci-build/smithy-rs-tool-common/src/package.rs:108]: There's a lot more to meta than just name and version (authors, license, etc). This should only ever be just the name and version.

File: tools/ci-build/publisher/src/cargo/publish.rs [hunk 2]
```diff
@@ -18,6 +20,10 @@ pub struct Publish {
 
 impl Publish {
     pub fn new(package_handle: PackageHandle, package_path: impl Into<PathBuf>) -> Publish {
+        assert!(
```
[the remark this answers is not in the document]
Comment (jdisanti) [reply tools/ci-build/publisher/src/cargo/publish.rs:23]: I tried it out, and while it works well for PackageHandle itself, it kind of breaks down for Package. Especially in the loading logic. I think it just makes things more complex to remove a few possible (yet easy to debug) panics.

**Pull request merged**